Utilities/StaticAnalyzers/src/MutableMemberModificationChecker.cpp

```diff
@@ -0,0 +1,205 @@
+    const auto *FD = llvm::dyn_cast<clang::FieldDecl>(ME->getMemberDecl());
+    if (!FD || !FD->isMutable()) {
+      return;  // Skip if it's not a mutable field
+    }
+
+    // == Skip non-private mutables, we deal with them elsewhere
+    if (FD->getAccess() != clang::AS_private) {
+      return;
+    }
+
+    // == Skip atmoic mutables, these are thread-safe by design ==
+    if (support::isStdAtomic(FD)) {
+      return;  // Skip if it's a mutable std::atomic
+    }
+
+    // == Check if a field is marked with special attributes ==
+    const clang::AttrVec &FAttrs = FD->getAttrs();
+    for (const auto *A : FAttrs) {
+      if (clang::isa<clang::CMSThreadGuardAttr>(A) || clang::isa<clang::CMSThreadSafeAttr>(A) ||
+          clang::isa<clang::CMSSaAllowAttr>(A)) {
+        return;
+      }
+    }
+
+    // == Check if we are inside a const-qualified member function ==
+    const auto *MethodDecl = llvm::dyn_cast<clang::CXXMethodDecl>(FuncD);
+    if (!MethodDecl || !MethodDecl->isConst()) {
+      return;
+    }
+
+    // == Check if we are modifying the mutable ==
+    bool ret;
+    ret = checkAssignToMutable(ME, C, FuncD);
+    if (!ret) {
+      ret = checkCallNonConstOfMutable(ME, C);
+    }
+
+    if (ret) {
+      if (RD) {
+        std::string ClassName = RD->getNameAsString();
+        std::string MemberName = ME->getMemberDecl()->getNameAsString();
+        std::string FunctionName = MethodDecl->getNameAsString();
+        std::string tname = "mutablemember-checker.txt.unsorted";
+        std::string ostring = "flagged class '" + ClassName + "' modifying mutable member '" + MemberName +
+                              "' in function '" + FunctionName + "'";
+        support::writeLog(ostring, tname);
+      }
+    }
+  }  // checkPreStmt
+
+  // Check direct modifications of mutable (assign, compound stmt, increment/decrement)
+  bool MutableMemberModificationChecker::checkAssignToMutable(const clang::MemberExpr *ME,
+                                                              clang::ento::CheckerContext &C,
+                                                              const clang::FunctionDecl *FuncD) const {
+    // == Check if this is a modifying statement ==
+    bool isModification = false;
+
+    // Retrieve the parent statement of the MemberExpr
+    const clang::LocationContext *LC = C.getLocationContext();
+    const clang::ParentMap &PM = LC->getParentMap();
+    const clang::Stmt *ParentStmt = PM.getParent(ME);
+
+    if (!ParentStmt) {
+      return false;
+    }
+
+    // Check if it is an assignment operator (binary operator)
+    const auto *BO = llvm::dyn_cast<clang::BinaryOperator>(ParentStmt);
+    if (BO) {
+      const auto *LHSAsME = llvm::dyn_cast<clang::MemberExpr>(BO->getLHS());
+      if (LHSAsME) {
+        if (BO->isAssignmentOp() && LHSAsME == ME) {
+          // The MemberExpr is on the left-hand side of an assignment
+          isModification = true;
```

Comment:
Would it make sense to `return false` directly here (and in the following checks) instead of setting the `isModification` boolean?